### ISSUE TITLE
Netlify example on SSR data?

### ISSUE THREAD

**Issue by skitsanos** [alibaba/ice#5624]
Guys,

I have a little issue here, probably because of the lack of documentation and examples. My deployment target is netlify, and actual rendering works just fine, you can see an example here: https://icejs-ssr.netlify.app/

For rendering, I am using this netlify function:

```ts
import {renderToHTML} from '../dist/server/index.mjs';
import {ServerContext} from '@ice/runtime';

const handler = async (event /*context*/) =>
{
    const {value} = await renderToHTML({
        req: event
    } as ServerContext);

    return {
        statusCode: 200,
        headers: {'Content-Type': 'text/html'},
        body: value
    };
};

export {handler};
```

What doesn't work for me on netlify is `serverDataLoader` bit:

```ts
import {defineServerDataLoader, useData} from 'ice';

export const serverDataLoader = defineServerDataLoader(async () =>
{
    return Promise.resolve({
        server: true,
        debug: 'yes'
        //data
    });
});

export default () =>
{
    const data = useData();

    return <div>Users page

        <pre>
                {JSON.stringify(data || {}, null, 2)}
                </pre>
    </div>;
}
```

But it does work fine when I do it via a basic HTTP server or on icejs dev server. 

The whole code you can check on https://github.com/skitsanos/icejs-ssr

Maybe anyone made it work on netlify and has a hint for me?

Thank you.

**Comment by chenjun1011** [comment]
May be the page previewed by netlify is not SSR, but SSG(html generated at build time). We're not familiar with netlify's release process right now, but will have a try later.

**Comment by skitsanos** [comment]
It seems I managed to make it work with adding `serverless-http`, so my render handler is like this now:

```ts
import serverless from 'serverless-http';
import {createServer} from 'http';
import {extname, join} from 'path';
import {createReadStream} from 'fs';
import {renderToHTML} from 'dist/server/index.mjs';

const server = createServer(async (req, res) =>
{
    const ext = extname(req.url);

    if (ext)
    {
        // static file url
        const path = join('dist', req.url);
        const stream = createReadStream(path);
        stream.on('error', (error) =>
        {
            res.writeHead(404, 'Not Found');
            res.end();
        });

        stream.pipe(res);
    }
    else
    {
        const {value} = await renderToHTML({
            req
        });

        res.write(value);
        res.end();
    }
});

const handler = serverless(server);

export {handler};
```

I can confirm, - it does work; I can see server-side data rendered into a page. But there is a trick:

> You need to remove from `dist`, where `ice build` saves output, all the HTML files generated.

That dataLoader thing is pretty confusing. if `defineServerDataLoader` is used, it will render the data only on document location change, but not on `history.push()` based page redirects, - to cover both of them I had to use `defineDataLoader`:

```ts
import {defineDataLoader,  useData} from 'ice';

export const dataLoader = defineDataLoader(async () =>
{
    const apiResponse = await fetch('https://api.skitsanos.com/api/utils/headers');

    return Promise.resolve({
        server: true,
        debug: 'yes',
        data: await apiResponse.json()
    });
});

export default () =>
{
    const data = useData();

    return <div>Users page

        <pre>
                {JSON.stringify(data || {}, null, 2)}
                </pre>
    </div>;
}
```

It does the trick, but I could spot that number of HTTP headers varies from the request done with `history.push()` and with a visiting page with ` document.location.href`. No so critical issue for me for now, but maybe someone can have a look into it as well.

Plus, if there is any way to disable HTML file generation via config, - it would be nice. Otherwise, this can be considered as *resolved*.

**Comment by chenjun1011** [comment]
Take the `users` page as an example:
- if it's visited directly by the url `/users`, it will be render in the server
- if it's pushed by `history.push()` , it will be render in the client

So, the code of `users`, will be run both in client & server.  

If data request code is isomorphic,  should use `dataLoader`.  If there are differences in the data request on the server side and the client side, can use `serverDataLoader` to write server only code.(We'll make that clear in the docs.)

In the `users` pages, it's better to use `dataLoader`, and use [isomorphic-unfetch](https://www.npmjs.com/package/isomorphic-unfetch) or [cross-fetch](https://github.com/lquixada/cross-fetch)  to fetch data. (but both isomorphic-unfetch and cross-fetch seems can not work probably in netlify, need to look again at the solution)

**Comment by chenjun1011** [comment]
Remove HTML file by the following config in `netlify.toml` may be a solution：

```toml
[build]
base = "."
publish = "./dist"
functions = "netlify-functions/"
command = "yarn run build & rm dist/*.html"
```

**Comment by chenjun1011** [comment]
>> but I could spot that number of HTTP headers varies from the request done with history.push() and with a visiting page with document.location.href

Which request and what's the differences? I may not understand the problem well.

**Comment by chenjun1011** [comment]
> Take the `users` page as an example:
> 
> * if it's visited directly by the url `/users`, it will be render in the server
> * if it's pushed by `history.push()` , it will be render in the client
> 
> So, the code of `users`, will be run both in client & server.
> 
> If data request code is isomorphic, should use `dataLoader`. If there are differences in the data request on the server side and the client side, can use `serverDataLoader` to write server only code.(We'll make that clear in the docs.)
> 
> It's better to use `dataLoader` in the `users` pages, and use [isomorphic-unfetch](https://www.npmjs.com/package/isomorphic-unfetch) or [cross-fetch](https://github.com/lquixada/cross-fetch) to fetch data. (But both isomorphic-unfetch and cross-fetch can not work probably after build in ice.js, it's a bug will be fixed #5626)

You can try @ice/app@3.0.1-alpha.1, fixed this issue.
